### ISSUE TITLE
Serve provenance without requiring the Accept header

### ISSUE THREAD

**Issue by simonw** [pypi/warehouse#17084]
**What's the problem this feature will solve?**

When I visit this page in my browser:

https://pypi.org/integrity/pydantic/1.10.19/pydantic-1.10.19-cp310-cp310-macosx_11_0_arm64.whl/provenance

I get this:

```json
{"message":"Request not acceptable"}
```
To see the content of that page I have to send an `accept` header like this:

```bash
curl -s \
  -H 'accept: application/vnd.pypi.integrity.v1+json' https://pypi.org/integrity/pydantic/1.10.19/pydantic-1.10.19-cp310-cp310-macosx_11_0_arm64.whl/provenance \
  | jq
```

**Describe the solution you'd like**

I'd prefer it if the page served me JSON without me having to send that accept header. That way I could explore and understand the API without needing to fire up a terminal or a custom HTTP client.

**Additional context**

Here's the implementation: https://github.com/pypi/warehouse/blob/06a2b586773dfeb72c604eb69396a2a8ede3f424/warehouse/api/integrity.py#L62-L67

**Comment by simonw** [comment]
This is covered by the documentation here (I didn't think to look so I dug around in the source code instead): https://docs.pypi.org/api/integrity/#get-provenance-for-file

**Comment by di** [comment]
Your browser is almost certainly sending `Accept: text/html` or something similar, which is not a bare `Accept` header. A bare `Accept` header correctly returns a JSON response:

```
$ curl -s 
    -H 'accept:' https://pypi.org/integrity/pydantic/1.10.19/pydantic-1.10.19-cp310-cp310-macosx_11_0_arm64.whl/provenance
    | jq

{
  "attestation_bundles": [
    {
       ...
```

**Comment by simonw** [comment]
Running `curl` without sending `-H 'accept:'` works too:
```bash
curl -s https://pypi.org/integrity/pydantic/1.10.19/pydantic-1.10.19-cp310-cp310-macosx_11_0_arm64.whl/provenance
```
Not great from a developer experience / usability POV though, since I do a lot of my API research these days on a phone with Mobile Safari!

**Comment by di** [comment]
> Running `curl` without sending `-H 'accept:'` works too:

That's because by default curl accepts any content type (`*/*`).

> Not great from a developer experience / usability POV though, since I do a lot of my API research these days on a phone with Mobile Safari!

The issue is that we want to respect the `Accept` header we are being sent: if your browser is indicating that it does not accept `application/json` for the request, then we don't want to send it `application/json`.

**Comment by SpecLad** [comment]
The `Accept` header handling seems to be broken.

Here I explicitly list the integrity JSON media type as acceptable, yet the server refuses to send it:

```console
$ curl -H 'Accept: text/html, application/vnd.pypi.integrity.v1+json; q=0.9' https://pypi.org/integrity/pydantic/1.10.19/pydantic-1.10.19-cp310-cp310-macosx_11_0_arm64.whl/provenance
{"message":"Request not acceptable"}
```

And here I don't list it, but the server _does_ send it:

```console
$ curl -H 'Accept: image/png' https://pypi.org/integrity/pydantic/1.10.19/pydantic-1.10.
19-cp310-cp310-macosx_11_0_arm64.whl/provenance
{"attestation_bundles":[{"attestations":[{"envelope":{"signature":"MEYCIQCwvxvlD0VNoopId6m0noBJHuJ/Ig6LPV/kckAb3c
...
```

Even when I expressly prohibit it, the server sends it anyway:

```console
$ curl -H 'Accept: image/png, application/vnd.pypi.integrity.v1+json; q=0' https://pypi.
org/integrity/pydantic/1.10.19/pydantic-1.10.19-cp310-cp310-macosx_11_0_arm64.whl/provenance
{"attestation_bundles":[{"attestations":[{"envelope":{"signature":"MEYCIQCwvxvlD0VNoopId6m0noBJHuJ/Ig6LPV/kckAb3c
...
```

**Comment by miketheman** [comment]
Not trying to solve this yet, but providing a little more context on why this behaves the current way.

The [code here](https://github.com/pypi/warehouse/blob/50a58f3081e693a3772c0283050a275e350004bf/warehouse/api/integrity.py#L28-L42) will take in the Accept header if provided, if none match, like the case of `image/png`, return the "default" content type of `application/vnd.pypi.integrity.v1+json`.

Then in the view, if any of the options passed [isn't the specific header, it's rejected](https://github.com/pypi/warehouse/blob/50a58f3081e693a3772c0283050a275e350004bf/warehouse/api/integrity.py#L66-L67).

So if no Accept header is provided, or an Accept header with something that isn't correct

This logic also denies the ability to pass the `q=0` option, as again, the logic in `_select_content_type()` will return an empty list, so the "default" fallback will be invoked.

This seems slightly incorrect from the [original intent to respond with a 404 if missing](https://github.com/pypi/warehouse/pull/16778#discussion_r1777448384).

**Comment by woodruffw** [comment]
I'll take a stab at improving the accept handling today 🙂

**Comment by woodruffw** [comment]
Oh, I see what's happening here in the `0.9` case:

With `Accept: text/html, application/vnd.pypi.integrity.v1+json; q=0.9` we end up selecting `text/html`, which _is_ currently listed as a valid offer (even though we reject it later), since it has the highest q-factor.

To fix this, I think we could probably tighten the accept handing: instead of doing the JSON fallback we could reject when no offer is accepted, and we could remove the HTML mimetypes from the offer list until we're actually ready to support them.

In other words, the behavior would become:

1. Unchanged: No `Accept:` header: assume they accept our default JSON form
2. Unchanged: `Accept: application/vnd.pypi.integrity.v1+json`: send our default JSON form
3. Everything else (including `q=0`): reject as unacceptable

Thoughts @miketheman @di?

**Comment by SpecLad** [comment]
Looks like this is now fixed - all of the test cases from the comments behave as expected.
